File: programmers/c++/가장먼노드.cpp
//가장 먼 노드
//bfs로 품
//최단거리 구할때
//check는 queue에 push할때
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;
const int MAX = 20001;
int dist[MAX];
bool check[MAX];
vector<int> graph[MAX];
int solution(int n, vector<vector<int>> edge) {
	int answer = 0;
	int maxLen = 0;
	queue<int> q;
	for (auto e : edge) {
		graph[e[0]].push_back(e[1]);
		graph[e[1]].push_back(e[0]);
	}
	q.push(1);
	dist[1] = 0;
	check[1] = true;
	while (!q.empty()) {
		int node = q.front();
		q.pop();
		vector<int> tmp = graph[node];
		for (int i = 0; i < tmp.size(); i++) {
			if (check[tmp[i]])
				continue;
			dist[tmp[i]] = dist[node] + 1;
			maxLen = max(maxLen, dist[tmp[i]]);
			check[tmp[i]] = true;
			q.push(tmp[i]);
		}
	}
	for (int i = 1; i <= n; i++) {
		if (dist[i] == maxLen)
			answer++;
	}
	return answer;
}
int main() {
	int n = 6;
	vector<vector<int>> edge = { {3, 6}, {4, 3}, {3, 2}, {1, 3}, {1, 2}, {2, 4}, {5, 2} };
	int ans = solution(n, edge);
	cout << ans;
	return 0;
}

File: programmers/python/graph/가장 먼 노드.py
#첫번째 풀이
#graph #인접리스트 #bfs
# 인접리스트를 이용해 그래프 생성 후 bfs로 각각의 최단거리 구해서 max값을 구함

from collections import deque

def solution(n, edge):
    answer = 0
    graph = [[] for _ in range(n + 1)]
    dist = [-1] * (n + 1)
    for vertex in edge:
        graph[vertex[0]].append(vertex[1])
        graph[vertex[1]].append(vertex[0])
    start = 1
    max_len = 0
    queue = deque([start])
    dist[start] = 0
    while queue:
        node = queue.popleft()
        for next_node in graph[node]:
            if dist[next_node] == -1:
                queue.append(next_node)
                dist[next_node] = dist[node] + 1
                max_len = dist[next_node]

    for node_dist in dist:
        if max_len == node_dist:
            answer += 1
    return answer

#두번째 풀이

from collections import deque

def solution(n, edge):
    answer = 0
    graph = [[] for _ in range(n + 1)]
    dist = [-1] * (n + 1)
    for (a , b) in edge: #(a,b)의 튜플로 받아서 더 깔끔한 코드로 수정
        graph[a].append(b)
        graph[b].append(a)
    start = 1
    max_len = 0
    queue = deque([start])
    dist[start] = 0
    while queue:
        node = queue.popleft()
        for next_node in graph[node]:
            if dist[next_node] == -1:
                queue.append(next_node)
                dist[next_node] = dist[node] + 1
                max_len = dist[next_node]
    
    answer = dist.count(max_len)
    #for문으로 하나씩 count증가안해도 count함수 사용하면 가능하다.

    return answer
